seq2struct/utils/parallelizer.py

```diff
@@ -0,0 +1,156 @@
+
+
+from abc import ABC, abstractmethod
+
+import time
+import queue
+import multiprocessing
+import tqdm
+
+class Parallelizer(ABC):
+    """
+    A parallelizer is a general purpose task used to handle the situation of
+        executing a function f(x, y) on a series of inputs x_i, y_i, where
+        x is often quite large and thus difficult to transport to a given
+        process, but there are typically fewer distinct values of x, allowing
+        for each process to use at most one value of x.
+
+    If it takes a constant amount of time `t_x` to ship x to a process and a constant
+        amount of time `t_y` to ship y to a process then execute f(x, y) and return the
+        result, then if we have `n` ys for a given x, then using `k` processes
+        takes time `k * t_x + n/k * t_y`, which is minimized when `t_x = n * t_y / k^2`,
+        or when `k = sqrt(n * t_y / t_x)`.
+
+    If `ratio=None` is passed in, we empirically `ratio = t_x/t_y`
+        with a 1-point estimate of `t_y` and `t_x`.
```

Comment:
For small jobs, it might be that the estimation takes more time than is saved later...

Reply:
changed it so does a 5 point estimate for greater accuracy and only does this if there are at least 50 data points. So at most a 10% increase in time consumption